File: ios/ViafouraModule.swift
import Foundation
import React

// Optional Viafoura SDK import. Code is guarded to build without it.
#if canImport(ViafouraSDK)
import ViafouraSDK

enum VFAdapterError: Error, LocalizedError {
  case invalidProvider
  case invalidInitialization
  case alreadyInitializedWithDifferentSite

  var errorDescription: String? {
    switch self {
    case .invalidProvider:
      return "Invalid social login provider"
    case .invalidInitialization:
      return "Invalid Viafoura initialization parameters"
    case .alreadyInitializedWithDifferentSite:
      return "ViafouraSDK is already initialized with a different site"
    }
  }
}

struct VFAuthAdapter {
  private let authService = ViafouraSDK.auth()

  func logout() {
    authService.logout()
  }

  func login(email: String, password: String, completion: @escaping (Result<Void, Error>) -> Void) {
    authService.login(email: email, password: password) { result in
      switch result {
      case .success:
        completion(.success(()))
      case .failure(let error):
        completion(.failure(error))
      }
    }
  }

  func signup(name: String, email: String, password: String, completion: @escaping (Result<Void, Error>) -> Void) {
    authService.signup(name: name, email: email, password: password, recaptchaToken: nil) { result in
      switch result {
      case .success:
        completion(.success(()))
      case .failure(let error):
        completion(.failure(error))
      }
    }
  }

  func socialLogin(token: String, provider: String, completion: @escaping (Result<Void, Error>) -> Void) {
    guard let p = VFSocialLoginProvider(rawValue: provider) else {
      completion(.failure(VFAdapterError.invalidProvider))
      return
    }
    authService.socialLogin(token: token, provider: p) { result in
      switch result {
      case .success:
        completion(.success(()))
      case .failure(let error):
        completion(.failure(error))
      }
    }
  }

  func loginRadiusLogin(token: String, provider: String?, completion: @escaping (Result<Void, Error>) -> Void) {
    let p: VFSocialLoginProvider
    if let provider, !provider.isEmpty {
      guard let parsed = VFSocialLoginProvider(rawValue: provider) else {
        completion(.failure(VFAdapterError.invalidProvider))
        return
      }
      p = parsed
    } else {
      p = .none
    }

    authService.loginRadiusLogin(token: token, provider: p) { result in
      switch result {
      case .success:
        completion(.success(()))
      case .failure(let error):
        completion(.failure(error))
      }
    }
  }

  func openIdLogin(token: String, completion: @escaping (Result<Void, Error>) -> Void) {
    authService.openIdLogin(token: token) { result in
      switch result {
      case .success:
        completion(.success(()))
      case .failure(let error):
        completion(.failure(error))
      }
    }
  }

  func cookieLogin(token: String, completion: @escaping (Result<Void, Error>) -> Void) {
    authService.cookieLogin(token: token) { result in
      switch result {
      case .success:
        completion(.success(()))
      case .failure(let error):
        completion(.failure(error))
      }
    }
  }

  func resetPassword(email: String, completion: @escaping (Result<Void, Error>) -> Void) {
    authService.resetPassword(email: email) { result in
      switch result {
      case .success:
        completion(.success(()))
      case .failure(let error):
        completion(.failure(error))
      }
    }
  }
}

struct VFCoreAdapter {
  private static let initLock = NSLock()
  private static var initializedKey: String?

  func initialize(siteUUID: String, siteDomain: String, enableLogging: Bool?) throws {
    guard let key = Self.initializationKey(siteUUID: siteUUID, siteDomain: siteDomain) else {
      throw VFAdapterError.invalidInitialization
    }

    Self.initLock.lock()
    defer { Self.initLock.unlock() }

    if let initializedKey = Self.initializedKey {
      guard initializedKey == key else {
        throw VFAdapterError.alreadyInitializedWithDifferentSite
      }
      if let enableLogging {
        ViafouraSDK.setLoggingEnabled(enableLogging)
      }
      return
    }

    if let enableLogging {
      ViafouraSDK.setLoggingEnabled(enableLogging)
    }
    ViafouraSDK.initialize(siteUUID: siteUUID, siteDomain: siteDomain)
    Self.initializedKey = key
  }

  private static func initializationKey(siteUUID: String, siteDomain: String) -> String? {
    let trimmedSiteUUID = siteUUID.trimmingCharacters(in: .whitespacesAndNewlines)
    let trimmedSiteDomain = siteDomain.trimmingCharacters(in: .whitespacesAndNewlines)

    guard
      let uuid = UUID(uuidString: trimmedSiteUUID),
      !trimmedSiteDomain.isEmpty
    else {
      return nil
    }

    return "\(uuid.uuidString.lowercased())::\(normalizeSiteDomain(trimmedSiteDomain))"
  }

  private static func normalizeSiteDomain(_ value: String) -> String {
    if let url = URL(string: value), let host = url.host, !host.isEmpty {
      return host.lowercased()
    }

    return value.lowercased()
  }
}

#else

enum VFAdapterError: Error, LocalizedError {
  case sdkUnavailable
  case invalidProvider
  case invalidInitialization
  case alreadyInitializedWithDifferentSite

  var errorDescription: String? {
    switch self {
    case .sdkUnavailable:
      return "ViafouraSDK is not installed. Add the iOS SDK to enable auth."
    case .invalidProvider:
      return "Invalid social login provider"
    case .invalidInitialization:
      return "Invalid Viafoura initialization parameters"
    case .alreadyInitializedWithDifferentSite:
      return "ViafouraSDK is already initialized with a different site"
    }
  }
}

struct VFAuthAdapter {
  func logout() { /* no-op when SDK unavailable */ }

  func login(email: String, password: String, completion: @escaping (Result<Void, Error>) -> Void) {
    completion(.failure(VFAdapterError.sdkUnavailable))
  }

  func signup(name: String, email: String, password: String, completion: @escaping (Result<Void, Error>) -> Void) {
    completion(.failure(VFAdapterError.sdkUnavailable))
  }

  func socialLogin(token: String, provider: String, completion: @escaping (Result<Void, Error>) -> Void) {
    completion(.failure(VFAdapterError.sdkUnavailable))
  }

  func loginRadiusLogin(token: String, provider: String?, completion: @escaping (Result<Void, Error>) -> Void) {
    completion(.failure(VFAdapterError.sdkUnavailable))
  }

  func openIdLogin(token: String, completion: @escaping (Result<Void, Error>) -> Void) {
    completion(.failure(VFAdapterError.sdkUnavailable))
  }

  func cookieLogin(token: String, completion: @escaping (Result<Void, Error>) -> Void) {
    completion(.failure(VFAdapterError.sdkUnavailable))
  }

  func resetPassword(email: String, completion: @escaping (Result<Void, Error>) -> Void) {
    completion(.failure(VFAdapterError.sdkUnavailable))
  }
}

struct VFCoreAdapter {
  func initialize(siteUUID: String, siteDomain: String, enableLogging: Bool?) throws {
    // no-op
  }
}

#endif


@objc(Viafoura)
public class ViafouraModule: NSObject {
  private let auth = VFAuthAdapter()
  private let core = VFCoreAdapter()

  @objc
  public static func requiresMainQueueSetup() -> Bool {
    return false
  }

  private func settle(
    _ result: Result<Void, Error>,
    _ resolve: @escaping RCTPromiseResolveBlock,
    _ reject: @escaping RCTPromiseRejectBlock
  ) {
    switch result {
    case .success:
      resolve(nil)
    case .failure(let error):
      reject("ERR_VIAFOURA", error.localizedDescription, error)
    }
  }

  @objc(logout:rejecter:)
  public func logout(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    auth.logout()
    resolve(nil)
  }

  @objc(login:password:resolver:rejecter:)
  public func login(
    _ email: String,
    password: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    auth.login(email: email, password: password) { self.settle($0, resolve, reject) }
  }

  @objc(signup:email:password:resolver:rejecter:)
  public func signup(
    _ name: String,
    email: String,
    password: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    auth.signup(name: name, email: email, password: password) { self.settle($0, resolve, reject) }
  }

  @objc(socialLogin:provider:resolver:rejecter:)
  public func socialLogin(
    _ token: String,
    provider: String?,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let provider = provider, !provider.isEmpty else {
      let error = VFAdapterError.invalidProvider
      reject("ERR_VIAFOURA", error.localizedDescription, error)
      return
    }
    auth.socialLogin(token: token, provider: provider) { self.settle($0, resolve, reject) }
  }

  @objc(loginRadiusLogin:provider:resolver:rejecter:)
  public func loginRadiusLogin(
    _ token: String,
    provider: String?,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    auth.loginRadiusLogin(token: token, provider: provider) { self.settle($0, resolve, reject) }
  }

  @objc(openIdLogin:resolver:rejecter:)
  public func openIdLogin(
    _ token: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    auth.openIdLogin(token: token) { self.settle($0, resolve, reject) }
  }

  @objc(cookieLogin:resolver:rejecter:)
  public func cookieLogin(
    _ token: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    auth.cookieLogin(token: token) { self.settle($0, resolve, reject) }
  }

  @objc(resetPassword:resolver:rejecter:)
  public func resetPassword(
    _ email: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    auth.resetPassword(email: email) { self.settle($0, resolve, reject) }
  }

  @objc(initialize:siteDomain:enableLogging:resolver:rejecter:)
  public func initialize(
    _ siteUUID: String,
    siteDomain: String,
    enableLogging: NSNumber?,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard !siteUUID.isEmpty, !siteDomain.isEmpty else {
      let error = VFAdapterError.invalidInitialization
      reject("ERR_VIAFOURA", error.localizedDescription, error)
      return
    }
    do {
      try core.initialize(
        siteUUID: siteUUID,
        siteDomain: siteDomain,
        enableLogging: enableLogging?.boolValue
      )
      resolve(nil)
    } catch {
      reject("ERR_VIAFOURA", error.localizedDescription, error)
    }
  }
}

File: android/src/main/java/com/viafoura/reactnative/ViafouraModule.kt
package com.viafoura.reactnative

import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import java.net.URL
import java.util.Locale
import java.util.UUID

import com.viafourasdk.src.ViafouraSDK
import com.viafourasdk.src.model.network.authentication.cookieLogin.CookieLoginResponse
import com.viafourasdk.src.model.network.authentication.login.LoginResponse
import com.viafourasdk.src.model.network.authentication.loginradius.LoginRadiusLoginResponse
import com.viafourasdk.src.model.network.authentication.openId.OpenIdLoginResponse
import com.viafourasdk.src.model.network.authentication.signup.SignUpResponse
import com.viafourasdk.src.model.network.authentication.socialLogin.SocialLoginResponse
import com.viafourasdk.src.model.network.error.NetworkError
import com.viafourasdk.src.services.auth.VFAuthService

class ViafouraModule(private val reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  companion object {
    const val NAME = "Viafoura"

    private val initLock = Any()
    private var initializedKey: String? = null

    private fun initializationKey(siteUUID: String, siteDomain: String): String? {
      val trimmedSiteUUID = siteUUID.trim()
      val trimmedSiteDomain = siteDomain.trim()

      if (trimmedSiteDomain.isBlank()) {
        return null
      }

      val normalizedUUID = try {
        UUID.fromString(trimmedSiteUUID).toString().lowercase(Locale.ROOT)
      } catch (e: IllegalArgumentException) {
        return null
      }

      return "$normalizedUUID::${normalizeSiteDomain(trimmedSiteDomain)}"
    }

    private fun normalizeSiteDomain(value: String): String {
      val host = try {
        URL(value).host
      } catch (e: Exception) {
        null
      }

      return (if (host.isNullOrBlank()) value else host).lowercase(Locale.ROOT)
    }
  }

  override fun getName(): String = NAME

  @ReactMethod
  fun logout(promise: Promise) {
    ViafouraSDK.auth().logout()
    promise.resolve(null)
  }

  @ReactMethod
  fun login(email: String, password: String, promise: Promise) {
    ViafouraSDK.auth().login(email, password, object : VFAuthService.LoginCallback {
      override fun onSuccess(loginResponse: LoginResponse) {
        promise.resolve(null)
      }
      override fun onError(err: NetworkError) {
        promise.reject("E_VF_LOGIN", err.message ?: "Login failed", null)
      }
    })
  }

  @ReactMethod
  fun signup(name: String, email: String, password: String, promise: Promise) {
    ViafouraSDK.auth().signup(name, email, password, object : VFAuthService.SignUpCallback {
      override fun onSuccess(loginResponse: SignUpResponse) {
        promise.resolve(null)
      }
      override fun onError(err: NetworkError) {
        promise.reject("E_VF_SIGNUP", err.message ?: "Signup failed", null)
      }
    })
  }

  @ReactMethod
  fun socialLogin(token: String, provider: String?, promise: Promise) {
    ViafouraSDK.auth().socialLogin(token, object : VFAuthService.SocialLoginCallback {
      override fun onSuccess(loginResponse: SocialLoginResponse) {
        promise.resolve(null)
      }
      override fun onError(err: NetworkError) {
        promise.reject("E_VF_SOCIAL", err.message ?: "Social login failed", null)
      }
    })
  }

  @ReactMethod
  fun loginRadiusLogin(token: String, provider: String?, promise: Promise) {
    ViafouraSDK.auth().loginRadiusLogin(token, object : VFAuthService.LoginRadiusLoginCallback {
      override fun onSuccess(loginResponse: LoginRadiusLoginResponse) {
        promise.resolve(null)
      }
      override fun onError(err: NetworkError) {
        promise.reject("E_VF_LOGINRADIUS", err.message ?: "LoginRadius login failed", null)
      }
    })
  }

  @ReactMethod
  fun openIdLogin(token: String, promise: Promise) {
    ViafouraSDK.auth().openIdLogin(token, object : VFAuthService.OpenIdLoginCallback {
      override fun onSuccess(loginResponse: OpenIdLoginResponse) {
        promise.resolve(null)
      }
      override fun onError(err: NetworkError) {
        promise.reject("E_VF_OPENID", err.message ?: "OpenID login failed", null)
      }
    })
  }

  @ReactMethod
  fun cookieLogin(token: String, promise: Promise) {
    ViafouraSDK.auth().cookieLogin(token, object : VFAuthService.CookieLoginCallback {
      override fun onSuccess(loginResponse: CookieLoginResponse) {
        promise.resolve(null)
      }
      override fun onError(err: NetworkError) {
        promise.reject("E_VF_COOKIE", err.message ?: "Cookie login failed", null)
      }
    })
  }

  @ReactMethod
  fun resetPassword(email: String, promise: Promise) {
    ViafouraSDK.auth().passwordReset(email, object : VFAuthService.PasswordResetCallback {
      override fun onSuccess() {
        promise.resolve(null)
      }
      override fun onError(err: NetworkError) {
        promise.reject("E_VF_RESET", err.message ?: "Password reset failed", null)
      }
    })
  }

  @ReactMethod
  fun initialize(
    siteUUID: String,
    siteDomain: String,
    enableLogging: Boolean?,
    promise: Promise
  ) {
    val context = reactContext.applicationContext
    if (context == null) {
      promise.reject("E_VF_INIT", "No application context available", null)
      return
    }
    if (siteUUID.isBlank() || siteDomain.isBlank()) {
      promise.reject("E_VF_INIT", "Invalid Viafoura initialization parameters", null)
      return
    }
    val initKey = initializationKey(siteUUID, siteDomain)
    if (initKey == null) {
      promise.reject("E_VF_INIT", "Invalid Viafoura initialization parameters", null)
      return
    }
    try {
      synchronized(initLock) {
        val currentKey = initializedKey
        if (currentKey != null) {
          if (currentKey != initKey) {
            throw IllegalStateException("ViafouraSDK is already initialized with a different site")
          }
          enableLogging?.let { ViafouraSDK.isLoggingEnabled = it }
          return@synchronized
        }

        enableLogging?.let { ViafouraSDK.isLoggingEnabled = it }
        ViafouraSDK.initialize(context, siteUUID, siteDomain)
        initializedKey = initKey
      }
      promise.resolve(null)
    } catch (e: Exception) {
      promise.reject("E_VF_INIT", e.message ?: "Viafoura initialization failed", e)
    }
  }
}
